### PR TITLE
Misc quality of life improvements

### PR DESCRIPTION
**Pull request by jmeggitt**
This pull request aims to address points of friction using the library. From my experience, those points of friction were:
 - Not many types implement `Hash`.
   - Makes it difficult to construct maps using these types as keys. Workarounds generally involve either making wrappers which can be hashed or converting types to primitives.
 - Long field chains and lots of nesting required to access values
   - Can be addressed by adding iterators and implementing `Deref` where appropriate.
 - A lot of boiler plate is required to access specific attributes
   - There are a few common attributes such as `AsPath` and `MpReachNlri` which are frequently used. Ideally, these attributes could be retrieved with helper methods instead of needing to iterate and match every attribute.
 - There is no easy way to map `*Value` enums to their respective `*Type` enums.
 - Error type is confusing to work with
   - Not covered by this pull request, but something worth looking into.
 - Types do not implement `Deserialize` (covered by #122).

## Notes on changes
 - `Attributes` is quite inefficient at the moment and has lots of areas for improvement. Firstly, it probably needs more methods for additional common attributes. I also chose to implement `IntoIterator` with `AttributeValue` since that is generally what the user is most interested in, but an argument can be made for using `Attribute` instead.
 - I think `AttrType` could have turned out better. Unfortunately, there is no default/other option for `#[derive(Primitive)]` so the inclusion of an unknown attribute type requires that it be implemented manually.
 - The `Attribute` struct adds additional boiler plate to accessing attributes. It would be easier to use if `AttributeValue` took the place of `Attribute` so it could be matched on directly. We can already determine the `AttrType` based on just `AttributeValue` so the only thing missing is the `Attribute` flags. Perhaps, the flags could be added to `AttributeValue` instead. Well-known attributes already have their flags pre-defined so they would only need to be added to some of the attributes. That being said, if new flags were to be added or the received flags are inconsistent with what is expected, they must be retained so they can be handled by the user.
 - I implemented `Debug` for a couple of types in an attempt to reduce the clutter when debug printing some values. I am not sure if this is worth keeping and may be better to simply derive.

### CODE REVIEW

File: src/models/bgp/attributes/aspath.rs [hunk 1]
```diff
@@ -31,7 +34,81 @@ impl AsPathSegment {
...
+        x.iter()
+            .sorted()
+            .zip(y.iter().sorted())
+            .all(|(a, b)| a == b)
+    }
+}
+
+impl Eq for AsPathSegment {}
+
+// TODO(jmeggitt): Implement iterators for AsPath (both over elements and all sequence variants)
```
Comment (digizeph) [inline src/models/bgp/attributes/aspath.rs:110]: This is a great idea 👍

File: src/models/bgp/attributes/mod.rs [hunk 2]
```diff
@@ -5,7 +5,12 @@ mod nlri;
 mod origin;
```
Comment (digizeph) [inline src/models/bgp/attributes/mod.rs:5]: A unit-test for the new attributes capabilities could be nice to add here, but I'm sure attributes will be tested elsewhere too, so not a big concern.

File: src/models/bgp/attributes/nlri.rs [hunk 3]
```diff
@@ -9,6 +17,59 @@ pub struct Nlri {
...
+
+    /// Returns true if this NLRI refers to reachable prefixes
+    pub const fn is_reachable(&self) -> bool {
+        self.next_hop.is_some()
+    }
+
+    /// Get the address of the next hop indicated by this NLRI.
+    ///
+    /// Panics if used on a unreachable NLRI message (ie. there is no next hop).
+    pub const fn addr(&self) -> IpAddr {
```
Comment (digizeph) [inline src/models/bgp/attributes/nlri.rs:39]: I'd suggest rename it to `nexthop_addr` or similar for clarity.

File: Cargo.toml [hunk 4]
```diff
@@ -27,6 +27,7 @@ serde={version="1.0", features=["derive"], optional = true}
 num-traits = {version = "0.2", optional = true}
 itertools = {version = "0.11.0", optional = true}
 ipnet = {version="2.7", optional = true}
+bitflags = {version="2.3.3", features = ["serde"], optional = true}
```
Comment (digizeph) [inline Cargo.toml:30]: `enum-primitive-derive` is no-longer needed, right?

File: src/models/bgp/community.rs [hunk 5]
```diff
@@ -112,6 +113,38 @@ pub enum ExtendedCommunity {
...
+                Some(ExtendedCommunityType::NonTransitiveIpv4AddressSpecific)
+            }
+            ExtendedCommunity::NonTransitiveFourOctetAsSpecific(_) => {
+                Some(ExtendedCommunityType::NonTransitiveFourOctetAsSpecific)
+            }
+            ExtendedCommunity::NonTransitiveOpaque(_) => {
+                Some(ExtendedCommunityType::NonTransitiveOpaque)
+            }
+            // TODO: Does Ipv6AddressSpecific not have a format type?
+            ExtendedCommunity::Ipv6AddressSpecific(_) => None,
```
Comment (digizeph) [inline src/models/bgp/community.rs:142]: I believe the IPv6 will have the same types as IPv4, but we will need to double check.

https://www.rfc-editor.org/rfc/rfc5701.html#section-2

```

    0                   1                   2                   3
    0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
   +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
   | 0x00 or 0x40  |    Sub-Type   |    Global Administrator       |
   +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
   |          Global Administrator (cont.)                         |
   +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
   |          Global Administrator (cont.)                         |
   +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
   |          Global Administrator (cont.)                         |
   +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
   | Global Administrator (cont.)  |    Local Administrator        |
   +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+

   The first high-order octet indicates whether a particular sub-type of
   this community is transitive across Autonomous Systems (ASes) (0x00),
   or not (0x40).  The second high-order octet of this extended type is
   used to indicate sub-types.  The sub-types are the same as for the
   IPv4 Address Specific Extended Community.
   ```

File: src/models/mrt/bgp4mp.rs [hunk 6]
```diff
@@ -18,28 +18,33 @@ pub enum BgpState {
...
+        match self {
+            Bgp4Mp::StateChange(x) => x.msg_type,
+            Bgp4Mp::Message(x) => x.msg_type,
+        }
+    }
 }
 
 /// BGP4MP message subtypes.
-#[derive(Debug, Primitive, Copy, Clone, PartialEq, Eq)]
+#[derive(Debug, Primitive, Copy, Clone, PartialEq, Eq, Hash)]
```
Comment (digizeph) [inline src/models/mrt/bgp4mp.rs:35]: Are we keeping the `Primitive` or should we also move the type ids to a separate field? (similar to what you did for attribute types)

File: src/models/mrt/mod.rs [hunk 7]
```diff
@@ -99,6 +100,7 @@ impl CommonHeader {
     }
 }
 
+// TODO: Why isn't there an option for an unknown message?
```
Comment (digizeph) [inline src/models/mrt/mod.rs:103]: Never encountered unknown message types before. The completeness, we may add a `Unknown` message, but it will probably not provide much value.

File: src/models/network/asn.rs [hunk 8]
```diff
@@ -1,28 +1,109 @@
...
+    /// within IANA's "Special-Purpose AS Numbers" registry. This includes checking against private
+    /// ASN ranges, ASNs reserved for documentation, and ASNs reserved for specific uses by various
+    /// RFCs.
+    ///
+    /// Up to date as of 2023-03-01 (Registry was last updated 2015-08-07).
+    ///
+    /// For additional details see:
+    ///  - <https://datatracker.ietf.org/doc/rfc7249/>
+    ///  - <https://www.iana.org/assignments/iana-as-numbers-special-registry/iana-as-numbers-special-registry.xhtml>
+    pub const fn is_public(&self) -> bool {
```
Comment (digizeph) [inline src/models/network/asn.rs:68]: I'd suggest we revert the code to have `is_reserved` function instead.

File: src/models/network/nexthop.rs [hunk 9]
```diff
@@ -1,17 +1,49 @@
...
+}
+
+// Attempt to reduce the size of the debug output
+impl Debug for NextHopAddress {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        match self {
+            NextHopAddress::Ipv4(x) => write!(f, "{}", x),
+            NextHopAddress::Ipv6(x) => write!(f, "{}", x),
+            // Is there a better notation for link local?
+            NextHopAddress::Ipv6LinkLocal(x, y) => write!(f, "Ipv6LinkLocal({}, {})", x, y),
```
Comment (digizeph) [inline src/models/network/nexthop.rs:42]: `Ipv6LinkLocal` debug looks good to me.

**Review by digizeph — COMMENTED** [review]
Great work! I left some comments for your questions.

File: Cargo.toml (hunk 4 above)
Comment (jmeggitt) [reply Cargo.toml:30]: I am thinking it may be better to replace it with the [`num_enum`](https://crates.io/crates/num_enum) crate. It does all of the same things as `enum-primitive-derive`, but is feature rich and appears to have more support. It supports deriving a catch-all variant which would remove the need to manually implement `From` the way I did for `AttrType`.

File: src/models/bgp/attributes/aspath.rs (hunk 1 above)
Comment (jmeggitt) [reply src/models/bgp/attributes/aspath.rs:110]: Sounds good, I'll add those to the pull request.

File: Cargo.toml (hunk 4 above)
Comment (digizeph) [reply Cargo.toml:30]: Sounds good! Let's replace it with num_enum

File: src/models/mrt/mod.rs (hunk 7 above)
Comment (jmeggitt) [reply src/models/mrt/mod.rs:104]: Yea, if we don't know the MRT type then there isn't exactly much that can even be parsed. I'm not sure what my reasoning was behind this particular comment.

File: src/models/bgp/error.rs
Comment (digizeph) [inline src/models/bgp/error.rs:1]: This is an excellent refactor!

File: src/models/network/afi.rs [hunk 10]
```diff
@@ -1,8 +1,11 @@
+use num_enum::{IntoPrimitive, TryFromPrimitive};
+
 /// AFI -- Address Family Identifier
 ///
 /// <https://www.iana.org/assignments/address-family-numbers/address-family-numbers.xhtml>
-#[derive(Debug, PartialEq, Primitive, Clone, Copy, Eq, Hash)]
+#[derive(Debug, PartialEq, TryFromPrimitive, IntoPrimitive, Clone, Copy, Eq, Hash)]
 #[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+#[repr(i16)]
```
Comment (digizeph) [inline src/models/network/afi.rs:8]: any reason this is `i16` instead of `u16`?

File: src/parser/bmp/error.rs [hunk 11]
```diff
@@ -31,14 +37,46 @@ impl Display for ParserBmpError {
 
 impl Error for ParserBmpError {}
 
-impl std::convert::From<std::io::Error> for ParserBmpError {
+// TODO: These conversions make the error difficult to debug as they drop all of the error context
```
Comment (digizeph) [inline src/parser/bmp/error.rs:40]: We could use [`thiserror`](https://docs.rs/thiserror/latest/thiserror/) crate for `ParserBmpError` or other errors. We could leave that for another PR if you prefer.

File: src/models/network/afi.rs (hunk 10 above)
Comment (jmeggitt) [reply src/models/network/afi.rs:8]: I used `u16` here because it was being converted to an `i16` before being parsed. I do not think any negative values have been assigned (other than 0xFFFF being reserved) so it should not make a difference either way. Do you think it would be better to use `u16`?

File: src/parser/bmp/error.rs (hunk 11 above)
Comment (jmeggitt) [reply src/parser/bmp/error.rs:40]: I think that would be a good idea. A separate PR may be better since any changes to the main error types will likely require widespread changes to the parser.

File: src/models/bgp/community.rs (hunk 5 above)
Comment (jmeggitt) [reply src/models/bgp/community.rs:144]: I read into it a bit, but it seems like IPv6 extended communities are actually a separate attribute. Perhaps it would be better to make it a separate struct and add it to `MetaCommunity`?

File: src/models/network/afi.rs (hunk 10 above)
Comment (digizeph) [reply src/models/network/afi.rs:8]: I think my original use of `i16` was a mistake. We should use `u16` for consistency.

File: src/models/bgp/community.rs (hunk 5 above)
Comment (digizeph) [reply src/models/bgp/community.rs:144]: I think it makes more sense to make it a `ExtendedCommunity` rather than a separate `MetaCommunity`.  We can probably use the original implementation

```rust
#[derive(Debug, PartialEq, Clone, Copy, Eq)]
#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
pub struct Ipv6AddressSpecific {
    pub ec_type: u8,
    pub ec_subtype: u8,
    // 16 octets
    pub global_administrator: Ipv6Addr,
    // 2 octets
    pub local_administrator: [u8; 2],
}
```

Or extend it a bit to make it `Transitive`/`NonTransitive` to match the IPv4 communities for symmetry.

File: src/models/bgp/attributes/aspath.rs [hunk 12]
```diff
@@ -31,30 +70,318 @@ impl AsPathSegment {
...
+
     pub fn from_segments(segments: Vec<AsPathSegment>) -> AsPath {
         AsPath { segments }
     }
 
-    pub fn add_segment(&mut self, segment: AsPathSegment) {
+    // TODO: Would it make more sense to prepend the segment instead of pushing to the end?
+    // Semantically, it feels like it would make more sense to get further from the origin when
+    // pushing a new segment.
+    pub fn push_segment(&mut self, segment: AsPathSegment) {
```
Comment (jmeggitt) [inline src/models/bgp/attributes/aspath.rs:306]: Should `push_segment` push the front or back of the `AsPath`? General convention would be to append to the back, but that would change the origin.

File: src/models/bgp/attributes/mod.rs [hunk 13]
```diff
@@ -109,15 +122,296 @@ pub fn get_deprecated_attr_type(attr_type: u8) -> Option<&'static str> {
...
+
+    pub fn origin_id(&self) -> Option<IpAddr> {
+        self.inner.iter().find_map(|x| match &x.value {
+            AttributeValue::OriginatorId(x) => Some(*x),
+            _ => None,
+        })
+    }
+
+    // TODO: Why would the attribute be used instead of the reach NLRI? Should this check both?
+    pub fn next_hop(&self) -> Option<IpAddr> {
```
Comment (jmeggitt) [inline src/models/bgp/attributes/mod.rs:163]: Would it be preferable to search for an `MP_REACHABLE_NLRI` attribute to get the next hop if the `NEXT_HOP` attribute is not found? From what I have seen, `NEXT_HOP` will sometimes not be included when `MP_REACHABLE_NLRI` is present despite being a mandatory attribute.

File: src/models/bgp/attributes/aspath.rs (hunk 12 above)
Comment (digizeph) [reply src/models/bgp/attributes/aspath.rs:306]: Why pushing back would change the origin? The origin is expected to be the last ASN of the AS path in this context. Combining with segments, it would be the last segment's last ASN (assuming the last segment is `AsSequence`). I don't want to change this fundamental ordering of the AS path unless necessary.

Thinking from another perspective: the first hop is always the closest to the route collector, and the origin of the prefix is the farthest.

Hope this makes sense.

File: src/models/bgp/attributes/mod.rs (hunk 13 above)
Comment (digizeph) [reply src/models/bgp/attributes/mod.rs:163]: I'd prefer for now to leave it as `next_hop` is an explicitly defined attribute and I don't want to do the smart thing and fill in the blank if it was mistakenly left empty. Unless there is a convincing legitimate case where `next_hop` should be added to the NLRI (or per-prefix next-hop in the same MRT record), I think we should just treat this as is.

Comment (jmeggitt) [reply src/models/bgp/attributes/mod.rs:163]: Sounds good.

For context though, I was curious about how frequently different attributes show up so I ran a quick test. In the test, I found that ~15% of `Attributes` within rib dumps did not contain `NEXT_HOP`.
```text
RIB Attribute Groups
COUNT         PERCENT NAME
[truncated]
77209439     1.40470% [ORIGIN, AS_PATH, NEXT_HOP, ATOMIC_AGGREGATE, COMMUNITIES]
82602566     1.50282% [ORIGIN, AS_PATH, NEXT_HOP, AGGREGATOR, COMMUNITIES]
83295574     1.51542% [ORIGIN, AS_PATH, NEXT_HOP, MULTI_EXIT_DISCRIMINATOR, COMMUNITIES, LARGE_COMMUNITIES]
96038091     1.74725% [ORIGIN, AS_PATH, NEXT_HOP, ATOMIC_AGGREGATE, AGGREGATOR, COMMUNITIES]
134907217    2.45441% [ORIGIN, AS_PATH, MULTI_EXIT_DISCRIMINATOR, COMMUNITIES, MP_REACHABLE_NLRI]
173909949    3.16400% [ORIGIN, AS_PATH, NEXT_HOP, MULTI_EXIT_DISCRIMINATOR]
212989898    3.87500% [ORIGIN, AS_PATH, MP_REACHABLE_NLRI]
241105973    4.38652% [ORIGIN, AS_PATH, NEXT_HOP, COMMUNITIES, LARGE_COMMUNITIES]
256900824    4.67388% [ORIGIN, AS_PATH, COMMUNITIES, MP_REACHABLE_NLRI]
737080900   13.40996% [ORIGIN, AS_PATH, NEXT_HOP, MULTI_EXIT_DISCRIMINATOR, COMMUNITIES]
1173907989  21.35731% [ORIGIN, AS_PATH, NEXT_HOP]
1564279355  28.45947% [ORIGIN, AS_PATH, NEXT_HOP, COMMUNITIES]

RIB Individual Attribute Occurrences
COUNT         PERCENT NAME
0            0.00000% AS4_PATH
0            0.00000% AIGP
0            0.00000% BGP_LS_ATTRIBUTE
0            0.00000% SFP_ATTRIBUTE
0            0.00000% PE_DISTINGUISHER_LABELS
0            0.00000% PMSI_TUNNEL
0            0.00000% BGPSEC_PATH
0            0.00000% BFD_DISCRIMINATOR
0            0.00000% TUNNEL_ENCAPSULATION
0            0.00000% TRAFFIC_ENGINEERING
0            0.00000% CLUSTER_ID
0            0.00000% ATTR_SET
0            0.00000% AS4_AGGREGATOR
0            0.00000% IPV6_ADDRESS_SPECIFIC_EXTENDED_COMMUNITIES
0            0.00000% MP_UNREACHABLE_NLRI
0            0.00000% BGP_PREFIX_SID
844          0.00002% Unknown(36)
1432         0.00003% Unknown(20)
1666         0.00003% Unknown(243)
5327         0.00010% Unknown(21)
405735       0.00738% DEVELOPMENT
838612       0.01526% ORIGINATOR_ID
838612       0.01526% CLUSTER_LIST
1043763      0.01899% LOCAL_PREFERENCE
17288435     0.31453% ONLY_TO_CUSTOMER
59600340     1.08433% EXTENDED_COMMUNITIES
446800968    8.12880% ATOMIC_AGGREGATE
523097119    9.51688% AGGREGATOR
591410005   10.75972% LARGE_COMMUNITIES
816150405   14.84850% MP_REACHABLE_NLRI
1336463411  24.31473% MULTI_EXIT_DISCRIMINATOR
3632860821  66.09388% COMMUNITIES
4680366060  85.15150% NEXT_HOP
5496516465 100.00000% AS_PATH
5496516465 100.00000% ORIGIN
```
I did find it strange though how `AS4_PATH` never occurred in the RIB dumps. I'm not sure if this indicates that something was broken during my changes.

The code and full data from that test is in this repo: https://github.com/jmeggitt/bgp_attribute_survey

Comment (digizeph) [reply src/models/bgp/attributes/mod.rs:163]: I think it's because `As4Path` is a value type that we created but not really used. In the parser function `parse_as_path`, we parse AS paths based on `asn_len` (i.e. 2byte or 4 byte) but will create a `AsPath` attribute value regardless of the length. I think we could probably remove the `As4Path` in the future.

Great work on the attribute survey! It is very interesting to learn about what attributes show up in the MRT files.

File: src/models/bgp/attributes/aspath.rs (hunk 12 above)
Comment (jmeggitt) [reply src/models/bgp/attributes/aspath.rs:306]: When I say back, I mean the back of the `Vec` of segments in memory, not the back of the path. As it is implemented currently, calling this function will change the origin of the path.

Lets say that I have a path with the segments `[AsSequence([3, 4, 5])]` and I call `push_segment(AsSequence([6]))`. What is the most intuitive outcome?
 - `AsPath { segments: [AsSequence([3, 4, 5]), AsSequence([6])] }` *(`Vec::push(x)` approach)*: This is the current implementation. After this call, the origin ASN changes from 5 to 6.
 - `AsPath { segments: [AsSequence([6]), AsSequence([3, 4, 5])] }` *(`Vec::insert(0, x)` approach)*: This does not change the origin and conceptually matches how a path gets extended as it gets farther from the origin while keeping the segments distinct.
 - `AsPath { segments: [AsSequence([3, 4, 5, 6])] }` *(merge push back approach)*: Same as the `Vec::push(x)` approach, but merges the segments together into a single sequence.
 - `AsPath { segments: [AsSequence([6, 3, 4, 5])] }` *(merge push front approach)*: Same as the `Vec::insert(0, x)` approach, but merges the segments together into a single sequence.

Comment (digizeph) [reply src/models/bgp/attributes/aspath.rs:306]: This probably falls into the very edge cases. If there were two `AsSequence` (which I doubt it happens in real life), the first scenario should be correct. It probably does not consider as "change of origin" if it is still in the process of parsing all segments. If the `AsSequnce([6])` comes after the other sequence in the MRT, it is closer to the origin. I don't think we should not attempt to merge two `AsSequence` if they appear. I think taking the last `AsSequence` as the AS path is safe.

Comment (jmeggitt) [reply src/models/bgp/attributes/aspath.rs]: Sounds good. I renamed the function to `append_segment` to hopefully make it a bit more clear and added some separate functions for coalescing and deduping segments of an `AsPath` for if that outcome is desired.

**Comment by jmeggitt** [comment]
Can the `ORIGINATOR_ID` and other BGP identifiers be IPv6 addresses? So far, I have only seen this attribute referenced as a 4 octet value corresponding to a BGP identifier.
https://github.com/bgpkit/bgpkit-parser/blob/e2d1ed4424207f100e508b5aeef4fc71bceb8c82/src/parser/bgp/attributes/attr_09_originator.rs#L10-L15

**Comment by digizeph** [comment]
> Can the `ORIGINATOR_ID` and other BGP identifiers be IPv6 addresses? So far, I have only seen this attribute referenced as a 4 octet value corresponding to a BGP identifier.
> 
> https://github.com/bgpkit/bgpkit-parser/blob/e2d1ed4424207f100e508b5aeef4fc71bceb8c82/src/parser/bgp/attributes/attr_09_originator.rs#L10-L15

They have to be 4 octets unless specified otherwise. They do not necessarily need to be IPv4 addresses. 

Corresponding [RFC4271 section](https://datatracker.ietf.org/doc/html/rfc4271#section-1.1) for `BGP Identifier`:

```

   BGP Identifier
      A 4-octet unsigned integer that indicates the BGP Identifier of
      the sender of BGP messages.  A given BGP speaker sets the value of
      its BGP Identifier to an IP address assigned to that BGP speaker.
      The value of the BGP Identifier is determined upon startup and is
      the same for every local interface and BGP peer.
```

Corresponding [RFC4456](https://www.rfc-editor.org/rfc/rfc4456.html#section-8) section for `ORIGINATOR_ID`:
```
   ORIGINATOR_ID is a new optional, non-transitive BGP attribute of Type
   code 9.  This attribute is 4 bytes long and it will be created by an
   RR in reflecting a route.  This attribute will carry the BGP
   Identifier of the originator of the route in the local AS.  A BGP
   speaker SHOULD NOT create an ORIGINATOR_ID attribute if one already
   exists.  A router that recognizes the ORIGINATOR_ID attribute SHOULD
   ignore a route received with its BGP Identifier as the ORIGINATOR_ID.
```

(when in doubt, check this attribute index: https://www.iana.org/assignments/bgp-parameters/bgp-parameters.xhtml#bgp-parameters-2)

File: src/models/bgp/attributes/mod.rs (hunk 2 above)
Comment (jmeggitt) [reply src/models/bgp/attributes/mod.rs:4]: Are there any specific changes that you want unit tests to cover?

**Comment by jmeggitt** [comment]
I think the changes are now about ready for review. I want to apologize again about the size of the PR. I did not realize it would get this large when I first started going off on tangents.

Please let me know which areas you feel require changes, additional unit test coverage, or should be deferred/removed. I imagine there are plenty of things I either forgot or overlooked.

Going forward, my current thinking is that the conversion to `thiserror` and any performance tweaks should be separate PRs.
